tensorflow/c/experimental/gradients/nn_grad_test.cc

```diff
@@ -125,6 +147,25 @@ class CppGradients
   bool UseFunction() const { return std::get<2>(GetParam()); }
 };
 
+TEST_P(CppGradients, TestReluGrad) {
+  // Mathematically, Relu isn't differentiable at `0`. So `gradient_checker`
+  // does not work with it.
+  float X_vals[] = {1.0f, 2.0f, 3.0f, -5.0f, -4.0f, -3.0f, 2.0f, 0.0f, -1.0f};
+  int64_t X_dims[] = {2, 2};
+  AbstractTensorHandlePtr X;
+  {
+    AbstractTensorHandle* X_raw;
+    Status s =
+        TestTensorHandleWithDimsFloat(ctx_.get(), X_vals, X_dims, 2, &X_raw);
+    ASSERT_EQ(errors::OK, s.code()) << s.error_message();
+    X.reset(X_raw);
+  }
+
+  ASSERT_NO_FATAL_FAILURE(CompareManualAndAutodiffGradients(
```

Comment:
Should we simply use `CompareNumericalAndAutodiffGradients` here? I think that is a more scalable way of adding tests for gradient functions.

Reply:
If we use it but we won't be albe to test the gradient computation at 0. WDYT ?

Reply:
I think in the "how to contribute" documents, we will encourage the use of `CompareNumericalAndAutodiffGradients`, `CompareManualAndAutodiffGradient` will only be used when `gradient_checker` does not work ( undifferentiable function, function with more than one outputs ( like MNISTForward )).

Reply:
Great point. Let's add both here then?

Another thing I was thinking, could we simplify `CompareNumericalAndAutodiffGradients` to only take the forward model fn? The gradient wrapper should be the same for all ops?

Reply:
> Let's add both here then?

Sorry I don't understand clearly what you means here :confused:

> Another thing I was thinking, could we simplify `CompareNumericalAndAutodiffGradients` to only take the forward model fn?

This is doable but it does not always work ( SpareSoftmax for example ). How about make `grad_model` an optional parameter ?

Reply:
I meant add 2 tests one with `CompareNumericalAndAutodiffGradients ` and other with `CompareManualAndAutodiffGradient ` to test grad at zero since we want all tests tested with `CompareNumericalAndAutodiffGradients`.

> This is doable but it does not always work ( SpareSoftmax for example ). How about make grad_model an optional parameter ?

Not sure I understand why this wont work for SparseSoftmax. Also I wasn't suggesting to make this change here. We could do this as a followup.

Reply:
> Not sure I understand why this wont work for SparseSoftmax. 

Because SparseSoftmax actually returns 2 outputs
https://github.com/tensorflow/tensorflow/blob/b8624d8c64c333b57aa1dc42c51af124cc9ddbb7/tensorflow/c/experimental/gradients/nn_grad_test.cc#L32-L42

In the `SparseSoftmaxCrossEntropyWithLogitsModel`, the second output is discarded because `gradient_checker` only needs the first output. But I am not sure if autodiff needs the second output or not. If it need, we will have to pass 2 different models.

Reply:
>  I meant add 2 tests

Done